basesandbox/src/execution/executor.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+
+use crate::ipc::semaphore::{Mode, Semaphore};
+use crate::ipc::Ipc;
+use std::process::Command;
+
+const SOCKET_SRC: &str = "./tmp/Foundry_Socket_Boxer";
+const SOCKET_DST: &str = "./tmp/Foundry_Socket_Boxee";
+const SEMAPHORE: &str = "/Foundry_Semaphore";
+
+struct Executor {
+    child: std::process::Child,
+}
+
+impl Executor {
+    fn new(path: &str, args: &[&str]) -> Self {
+        let mut command = Command::new(path);
+        command.args(args);
+        Executor {
+            child: command.spawn().unwrap(),
+        }
+    }
+}
+
+impl Drop for Executor {
+    fn drop(&mut self) {
+        // This is synchronized with excutee's signal,
+        // which is supposed to be sent in termination step.
+        // Thus, in normal case, it won't take much time to be in a waitable status.
+        // However a malicous excutee might send a signal arbitrarily before the termination.
+        // For that, we have a timeout for the wait.
+        for _ in 0..10 {
+            std::thread::sleep(std::time::Duration::from_millis(50));
+            if self.child.try_wait().unwrap().is_some() {
+                return // successful
+            }
+        }
+        panic!("Module hasn't terminated itself. Protocol Error")
+    }
+}
+
+struct DirectoryReserver {
+    path: String,
+}
+
+impl DirectoryReserver {
+    fn new(path: String) -> Self {
+        std::fs::create_dir(&path).unwrap();
+        DirectoryReserver {
+            path,
+        }
+    }
+}
+
+impl Drop for DirectoryReserver {
+    fn drop(&mut self) {
+        std::fs::remove_dir(&self.path).unwrap();
+    }
+}
+
+/// declaration order of fields is important because of Drop dependencies
+pub struct Context<T: Ipc> {
+    pub ipc: T,
+    pub semaphore: Semaphore,
+    _child: Executor,
+    _directory: DirectoryReserver,
+}
+
+// Note that field's drop is called later that of struct.
+impl<T: Ipc> Drop for Context<T> {
+    fn drop(&mut self) {
+        self.semaphore.wait();
+    }
+}
+
+/// id must be unique for each instance.
+pub fn execute<T: Ipc>(path: &str, id: &str) -> Result<Context<T>, String> {
```

Comment:
The fixed string id disallows you to run multiple foundry nodes in the same machine.
Change it to u128 and pass the timestamp prefixed random number.

Reply:
I'll use `H256`. In next PR, I'll add`socket.connect()` not to pass long socket address every time we `send()`.

Reply:
I found that using a process id and a module id make the name unique, and it doesn't need to hash.
```rust
let pid = std::process::id();
let adddr_src = format!("{}.{}.{}", SOCKET_SRC, pid, id);
```